components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java

```diff
@@ -1422,6 +1422,51 @@ public Subscriber getSubscriber(String subscriberName) throws APIManagementExcep
         return subscribedAPIs;
     }
 
+    public Set<Scope> getScopesForApplicationSubscription(Subscriber subscriber, int applicationId)
+            throws APIManagementException {
+        Connection connection = null;
+        ResultSet resultSet = null;
+        PreparedStatement ps = null;
+        HashMap<String, Scope> scopeHashMap = new HashMap<>();
+        int tenantId = APIUtil.getTenantId(subscriber.getName());
+
+        try (Connection conn = APIMgtDBUtil.getConnection()) {
+            String sqlQueryforGetSubscribedApis = SQLConstants.GET_SUBSCRIBED_API_IDs_BY_APP_ID_SQL;
+            String sqlQuery = SQLConstants.GET_SCOPE_BY_SUBSCRIBED_API_PREFIX + sqlQueryforGetSubscribedApis +
+                    SQLConstants.GET_SCOPE_BY_SUBSCRIBED_ID_SUFFIX;
+
+            if (conn.getMetaData().getDriverName().contains("Oracle")) {
+                sqlQuery = SQLConstants.GET_SCOPE_BY_SUBSCRIBED_ID_ORACLE_SQL + sqlQueryforGetSubscribedApis +
+                        SQLConstants.GET_SCOPE_BY_SUBSCRIBED_ID_SUFFIX;
+            }
+            try (PreparedStatement statement = conn.prepareStatement(sqlQuery)) {
+                statement.setInt(1, tenantId);
+                statement.setInt(2, applicationId);
+                try (ResultSet result = statement.executeQuery()) {
+                    while (resultSet.next()) {
+                        Scope scope;
+                        String scopeKey = resultSet.getString(1);
+                        if (scopeHashMap.containsKey(scopeKey)) {
+                            // scope already exists append roles.
+                            scope = scopeHashMap.get(scopeKey);
+                            scope.setRoles(scope.getRoles().concat("," + resultSet.getString(4)).trim());
+                        } else {
+                            scope = new Scope();
+                            scope.setKey(scopeKey);
+                            scope.setName(resultSet.getString(2));
+                            scope.setDescription(resultSet.getString(3));
+                            scope.setRoles(resultSet.getString(4).trim());
+                        }
+                        scopeHashMap.put(scopeKey, scope);
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            handleException("Failed to retrieve scopes ", e);
```

Comment:
Can we make the error message more descriptive?

Reply:
Fixed with a61e8a6c507360351ecc5afaefe315144054c52e